### PR TITLE
Feature/account management

### PR DESCRIPTION
**Pull request by Akeron44**

### CODE REVIEW

File: src/modules/account/account.controller.spec.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import { Test, TestingModule } from '@nestjs/testing';
+import { AccountController } from './account.controller';
+
+describe('AccountController', () => {
+  let controller: AccountController;
+
+  beforeEach(async () => {
```
Comment (zeroabsolute) [inline src/modules/account/account.controller.spec.ts:7]: @Akeron44 We need more tests

File: src/modules/account/account.service.ts [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+import { UpdateAccountDto } from './dtos/update-account.dto';
+import { ExceptionsHandler } from '@nestjs/core/exceptions/exceptions-handler';
+
+@Injectable()
+export class AccountService {
+  constructor(private prisma: PrismaService) {}
+
+  async createAccount(data: CreateAccountDto, userId: number) {
+    try {
+      if (!userId) return null;
```
Comment (zeroabsolute) [inline src/modules/account/account.service.ts]: @Akeron44 If we don't have a user id, something is wrong. We should throw an exception instead

File: src/modules/account/account.service.ts [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+      if (Object.keys(data).length === 0) {
+        throw new BadRequestException('No data provided.');
```
Comment (zeroabsolute) [inline src/modules/account/account.service.ts]: @Akeron44 Input/Request validation could be moved to a separate file just for clarity

File: src/modules/account/account.service.ts [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+      const account = await this.prisma.account.create({
+        data: accountInformation,
+      });
+
+      const role = await this.prisma.role.findFirst({
+        where: { roleName: UserRoles.ADMIN },
+      });
+
+      await this.prisma.membership.create({
```
Comment (zeroabsolute) [inline src/modules/account/account.service.ts]: @Akeron44 What is membership creation fails? We must make sure that all or nothing gets created. 
https://www.prisma.io/docs/orm/prisma-client/queries/transactions

File: src/modules/account/account.service.ts [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+          userId,
+          accountId: account.id,
+          roleId: role.id,
+          isConfirmed: true,
+        },
+      });
+
+      return account;
+    } catch (error) {
+      return new ExceptionsHandler(error.response);
```
Comment (zeroabsolute) [inline src/modules/account/account.service.ts]: @Akeron44 return new ExceptionHandler -> I would suggest a different name that would describe the error exactly.
And like discussed, the service could just throw the errors. Controller can decide what to do with them (e.g. return the correct http response payload and status based on the error type)

File: src/modules/account/account.service.ts [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        where: {
+          accountId,
+          userId,
+          isConfirmed: true,
+          deletedAt: null,
+        },
+      });
+
+      if (!membership) {
+        throw new BadRequestException('Bad request.');
```
Comment (zeroabsolute) [inline src/modules/account/account.service.ts]: @Akeron44 I don't think this is the correct error. Instead we could throw ResourceNotFound. The request is not malformed, we just did not find any records

File: src/modules/account/account.service.ts [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+      if (Object.keys(newAccountInformation).length === 0) {
+        throw new BadRequestException('No data provided.');
+      }
+
+      const currentAccount = await this.prisma.account.findFirst({
+        where: { id: accountId, deletedAt: null },
+      });
+
+      if (!currentAccount) {
+        throw new BadRequestException('Bad request.');
```
Comment (zeroabsolute) [inline src/modules/account/account.service.ts]: @Akeron44 Same as above, it's not a bad request

File: src/modules/account/account.service.ts [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+        data: { ...newAccountInformation },
+      });
+    } catch (error) {
+      return new ExceptionsHandler(error.response);
+    }
+  }
+
+  async deleteAccount(accountId: number) {
+    try {
+      if (!accountId) return null;
```
Comment (zeroabsolute) [inline src/modules/account/account.service.ts]: @Akeron44 Throw an error instead of returning null

File: src/modules/account/account.service.ts [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+      const memberships = await this.prisma.membership.findMany({
+        where: { accountId, isConfirmed: true, deletedAt: null },
+      });
+
+      if (memberships.length === 0) {
+        return memberships;
+      }
+
+      await this.prisma.membership.deleteMany({
```
Comment (zeroabsolute) [inline src/modules/account/account.service.ts]: @Akeron44 Again transaction needed to make sure we removed all related resources

File: src/modules/membership/membership.controller.spec.ts [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+  beforeEach(async () => {
+    const module: TestingModule = await Test.createTestingModule({
+      controllers: [MembershipController],
+    }).compile();
+
+    controller = module.get<MembershipController>(MembershipController);
+  });
+
+  it('should be defined', () => {
```
Comment (zeroabsolute) [inline src/modules/membership/membership.controller.spec.ts:15]: @Akeron44 More tests needed

File: src/modules/membership/membership.service.ts [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+          role: {
+            select: {
+              roleName: true,
+            },
+          },
+        },
+      });
+
+      if (!membership) {
+        throw new BadRequestException('Bad request.');
```
Comment (zeroabsolute) [inline src/modules/membership/membership.service.ts]: @Akeron44 Same, wrong error thrown

**Comment by zeroabsolute** [comment]
@Akeron44 There are also conflicts with the master branch that need to be solved

File: src/main.ts [hunk 12]
```diff
@@ -1,33 +1,84 @@
...
+      'https://cdnjs.cloudflare.com/ajax/libs/swagger-ui/4.15.5/swagger-ui-standalone-preset.min.css',
+      'https://cdnjs.cloudflare.com/ajax/libs/swagger-ui/4.15.5/swagger-ui.css',
+    ],
+  });
+  await app.listen(process.env.PORT || 8000);
 }
+
 bootstrap();
+
+// import { NestFactory } from '@nestjs/core';
```
Comment (gerimujo496) [inline src/main.ts]: remove the comment

File: src/modules/account/account.service.ts [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+import { validateNonEmptyObject } from 'src/common/util/validator';
+
+@Injectable()
+export class AccountService {
+  constructor(private accountDAL: AccountDAL) {}
+
+  async createAccount(data: CreateAccountDto, userId: number) {
+    if (!userId) throw new BadRequestException('User ID is required.');
+
+    const user = await this.accountDAL.findUserById(userId);
```
Comment (gerimujo496) [inline src/modules/account/account.service.ts:21]: i have created a function in my auth controller to get the user or throw error , you can use that

File: src/modules/account/account.service.ts [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+
+    const [account, membership] =
+      await this.accountDAL.createMembershipAndAccount(data, userId, password);
+
+    return { account, membership };
+  }
+
+  async getMyAccount(userId: number, accountId: number) {
+    if (!userId || !accountId)
+      throw new BadRequestException(
```
Comment (gerimujo496) [inline src/modules/account/account.service.ts:40]: for errors please use constants

File: src/modules/account/account.service.ts [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+        'User ID and Account ID are both required.',
+      );
+
+    const account = await this.accountDAL.findAccount(accountId);
+    await this.accountDAL.findMembership(accountId, userId);
+
+    return account;
+  }
+
+  async getAccountUsers(accountId: number) {
```
Comment (gerimujo496) [inline src/modules/account/account.service.ts:48]: and here use a constant

File: src/modules/account/account.service.ts [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+
+    return await this.accountDAL.findUsersMembershipsByAccount(accountId);
+  }
+
+  async updateAccount(
+    accountId: number,
+    newAccountInformation: UpdateAccountDto,
+  ) {
+    if (!accountId) throw new BadRequestException('Account ID is required.');
+
```
Comment (gerimujo496) [inline src/modules/account/account.service.ts:72]: and here use a constant

File: src/modules/membership/dal/membership.dal.ts [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+        'Invalid membership or membership not found.',
+      );
+    }
+
+    return membership;
+  }
+
+  async findNotAdminMembership(accountId: number, userId: number) {
+    try {
+      const notAdminMembership = await this.prisma.membership.findFirst({
```
Comment (gerimujo496) [inline src/modules/membership/dal/membership.dal.ts:36]: please no business logic in dal

File: src/modules/membership/membership.controller.ts [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+} from '@nestjs/common';
+import { UpdateRoleDto } from './dtos/update-role.dto';
+import { MembershipService } from './membership.service';
+import { AuthGuard } from 'src/common/guards/auth.guard';
+import { Roles } from 'src/common/decorators/roles.decorator';
+import { UserRoles } from '@prisma/client';
+import { RolesGuard } from 'src/common/guards/roles.guard';
+import { controller_path } from 'src/constants/controller-path';
+
+@Controller('membership')
```
Comment (gerimujo496) [inline src/modules/membership/membership.controller.ts]: and here use a const

File: src/modules/membership/membership.service.ts [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+import { MembershipDAL } from './dal/membership.dal';
+
+@Injectable()
+export class MembershipService {
+  constructor(private membershipDAL: MembershipDAL) {}
+
+  async updateUserRole({ accountId, userId, role }: UpdateMembership) {
+    if (!userId || !accountId)
+      throw new BadRequestException(
+        'User ID and Account ID are both required.',
```
Comment (gerimujo496) [inline src/modules/membership/membership.service.ts]: use a const

File: src/modules/membership/membership.service.ts [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+        'Membership ID and Account ID are both required.',
+      );
+
+    const membership = await this.membershipDAL.findMembership(
+      membershipId,
+      accountId,
+    );
+
+    if (membership.role.roleName === UserRoles.ADMIN) {
+      throw new ForbiddenException('Forbidden access.');
```
Comment (gerimujo496) [inline src/modules/membership/membership.service.ts]: use a const

File: src/modules/membership/dal/membership.dal.ts (hunk 17 above)
Comment (Akeron44) [reply src/modules/membership/dal/membership.dal.ts:36]: This is not business logic, it is error handling

**Review by gerimujo496 — APPROVED** [review]

**Review by danielaKokoneshi309 — APPROVED** [review]

**Pull request merged**